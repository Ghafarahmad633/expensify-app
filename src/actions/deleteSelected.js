export const addItemToDelete=({id})=>({
    type:"PUSH_EXPENSE_TO_DELETE",
    id
})
export const removeItemToDelete=({id})=>({
    type:"REMOVE_EXPENSE_TO_DELETE",
    id
})
export const refereshStateAfterDelete=()=>({
    type:"REFRESH_STATE_AFTER_DELETE",
})
export const setTrueSelectAll=()=>({
    type:"SET_TRUE_SELECT_ALL"
})
export const setFalseSelectAll=()=>({
    type:"SET_FALSE_SELECT_ALL"
})