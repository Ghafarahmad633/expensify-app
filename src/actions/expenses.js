import uuid from "uuid";
import database from '../firebase/firebase'
// this is ADD_EXPENSE

export const addExpenses=(expense)=>({
    type:"ADD_EXPENSE",
    expense

});

export const startAddExpense=(expenseData={})=>{
    return (dispatch,getState)=>{
        const uid=getState().auth.uid;
        const {
            desc='',
            note='',
            createdAt=0,
            amount=0
        }=expenseData
        const expense={desc,note,createdAt,amount}
        return database.ref(`users/${uid}/expenses`).push(expense)
            .then((ref)=>{
                dispatch(addExpenses({
                    id:ref.key,
                    ...expense
                }))
            })

    }
}

// this is REMOVE_EXPENSE
export const removeExpense=({id}={})=>({
    type: "REMOVE_EXPENSE",
    id

})
export const startRemoveExpense=({id})=>{
    return (dispatch,getState)=>{
        const uid=getState().auth.uid
        return database.ref(`users/${uid}/expenses/${id}`).remove()
            .then(()=>{
                dispatch(removeExpense({id}))
            })
    }
}
export const removeExpenseSelected=(id)=>({
    type:"DELETE_SELECTED",
    id
})
// This is editExpense
export const editExpense=(id,updates)=>({
    type:"EDIT_EXPENSE",
    id,
    updates

})
export const startEditExpense=(id,updates)=>{
    return (dispatch,getState)=>{
        const uid=getState().auth.uid
       return database.ref(`users/${uid}/expenses/${id}`).update({
            ...updates
        }).then(()=>{
            dispatch(editExpense(id,updates))
       })
    }
}
export const removeExpenseSelctedRange=({ids})=>({
    type:"REMOVE_EXPENSES_SELECTED_RANGE",
    ids

})

export const startRemoveExpenseSelctedRange=({ids})=>{
    return (dispatch,getState)=>{
        const uid=getState().auth.uid
        const objectToIds={}
        ids.forEach((id)=>{
            objectToIds[id]=null
        })
        return database.ref(`users/${uid}/expenses`).update(objectToIds).then(()=>{
            dispatch(removeExpenseSelctedRange({ids}))
        })
    }
}

export const setExpense=(expenses)=>({
    type:"SET_EXPENSES",
    expenses
})
export const startSetExpense=()=>{
    return(dispatch,getState)=>{
        const uid=getState().auth.uid
       return database.ref(`users/${uid}/expenses`).once('value')
            .then((snapshot)=>{
                const expenses=[]
                snapshot.forEach((childSnaphot)=>{
                    expenses.push({
                        id:childSnaphot.key,
                        ...childSnaphot.val()
                    })
                })
                dispatch(setExpense(expenses))
            })

    }
}