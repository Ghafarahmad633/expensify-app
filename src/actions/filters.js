
// Filter Reducer
export const filterDefault={
    text:'',
    sortBy:'',
    startDate:undefined,
    endDate:undefined
};
// This is SET_TEXT_FILTER
export const setFiltertext=(text='')=>({
    type:"SET_TEXT_FILTER",
    text

});
// this is SORT_BY_DATE
export const sortyByDate=()=>({
    type:"SORT_BY_DATE",

});
// this is SORT_BY_AMOUNT
export const sortyByAmount=()=>({
    type:"SORT_BY_AMOUNT",

});
// this is SET_START_DATE
export const setStartDate=(date)=>({
    type:"SET_START_DATE",
    date

});
// this is SET_END_DATE
export const setEndDate=(date)=>({
    type:"SET_END_DATE",
    date

});