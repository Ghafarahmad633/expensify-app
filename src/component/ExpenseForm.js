import React from 'react'
import moment from 'moment'
import {SingleDatePicker} from 'react-dates'

const now=moment();
console.log(now.format('MMM,Do, YYYY'))
class ExpenseForm extends React.Component{
    constructor(props){
        super(props);
        this.state={
            decs: props.expense?props.expense.desc:"",
            note:props.expense?props.expense.note:"",
            amount:props.expense?(props.expense.amount/100).toString():"",
            createdAt:props.expense?moment(props.expense.createdAt):moment(),
            calendarFocused:false,
            error:undefined,
            buttonText:props.buttonText?props.buttonText:"Submit"
        }
    }
    onChangeDecs=(e)=>{
        const decs=e.target.value
        this.setState(()=>({decs}))
    };
    onNoteChange=(e)=>{

        const note=e.target.value
        this.setState(()=>({note}))

    };
    onAmountChange=(e)=>{

        const amount=e.target.value;
        if(!amount || amount.match(/^\d{1,}(\.\d{0,2})?$/)){
            this.setState(()=>({amount}))

        }
    };
    onDateChange=(createdAt)=>{

       if(createdAt){
           this.setState(()=>({createdAt}))
       }

    };
    onFocusChange=({focused})=>{
        this.setState(()=>({calendarFocused: focused}))
    };
    onSubmit=(e)=>{
        e.preventDefault()
        if(!this.state.decs||!this.state.amount){
            this.setState(()=>({error: "Please Fill All Fields"}))
            console.log("Some Error")
        }
        else {
            this.setState(()=>({error: ''}))
            this.props.onSubmit(({
                desc:this.state.decs,
                amount:parseFloat(this.state.amount,10)*100,
                createdAt:this.state.createdAt.valueOf(),
                note:this.state.note

            }))
        }
    }
    render(){
        return(
            <div>
                <form onSubmit={this.onSubmit}>
                    <input
                        name='desc'
                        type='text'
                        placeholder='Decritpion'
                        onChange={this.onChangeDecs}
                        value={this.state.decs}
                    />
                    <input
                        type='text'
                        name='amount'
                        placeholder='Amount'
                        value={this.state.amount}
                        onChange={this.onAmountChange}
                    />
                    <SingleDatePicker
                        date={this.state.createdAt}
                        onDateChange={this.onDateChange}
                        focused={this.state.calendarFocused}
                        onFocusChange={this.onFocusChange}
                        numberOfMonths={1}
                        isOutsideRange={(day)=>false}
                    />
                    <textarea
                        name='note'
                        onChange={this.onNoteChange}
                        value={this.state.note}
                        placeholder='add your note'
                    >

                    </textarea>
                    <button type='submit'>{this.state.buttonText}</button>
                </form>
                {this.state.error!==undefined&&<p>Please Fill out the required fields</p>}
            </div>
        )
    }

}
export default ExpenseForm