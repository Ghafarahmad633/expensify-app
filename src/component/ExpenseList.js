import React from 'react'
import {connect} from 'react-redux'
import ExpenseListItem from '../component/ExpenseListItem'
import selcectExpense from '../selectors/expenses'
import {addItemToDelete,removeItemToDelete} from '../actions/deleteSelected'
export class ExpenseList extends React.Component{
    constructor(props){
        super(props)
    };
    state={
        selectAll:this.props.selectAllExpense===true?'checked':null
    }
    onSelectChange=(e)=>{
        if(e.target.checked){
            this.props.addItemToDelete({id:e.target.value})

        }
        else if(!e.target.checked){
            this.props.removeItemToDelete({id:e.target.value})
        }



    };
    render(){
        return(
            <div className='content-continer'>
                <div className='list-header'>
                    <div className='show-for-mobile'>Expenses</div>
                    <div className='show-for-desktop'>Expense</div>
                    <div className='show-for-desktop'>Amount</div>
                </div>
               <div className='list-item__body'>
                   {
                       this.props.expenses.length===0?(
                           <div className='list-header__message'>No Expenses Found</div>
                       ):(
                           this.props.expenses.map((expenses)=>(
                               <ExpenseListItem
                                   key={expenses.id}
                                   {...expenses}
                                   onSelectChange={this.onSelectChange}
                                   selectAll={this.state.selectAll}
                               />
                           ))
                       )
                   }
               </div>

            </div>
        )
    }
}

const mapStatProps=(state)=>{
    return{
        expenses:selcectExpense(state.expense,state.filters),
        expenseToDelete:state.selectedToDeleted,
        selectAllExpense:state.selectAllExpense

    }
}
const mapDispatchToProps=(dispatch,props)=>{
    return{
        addItemToDelete:(data)=>dispatch(addItemToDelete(data)),
        removeItemToDelete:(data)=>dispatch(removeItemToDelete(data))
    }
}
export default connect(mapStatProps,mapDispatchToProps)(ExpenseList)