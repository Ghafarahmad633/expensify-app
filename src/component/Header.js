import React from "react";
import {connect} from 'react-redux'
import {Link} from "react-router-dom";

import {startLogOut} from '../actions/auth';


export const Header=({startLogOut,isUserLogin})=>(
    <div className='header'>
        <div className='content-continer'>

            <div className='header__content'>
                <Link className='header__title'  to="/dashboard"><h1>Expensifiy</h1></Link>
                <button className='button button__link' onClick={startLogOut}>Logout</button>
            </div>
        </div>

    </div>
)
const mapDispatchToProps=(dispatch)=>{
    return{
        startLogOut:()=>dispatch(startLogOut()),
    }
}

export default connect(undefined,mapDispatchToProps)(Header)