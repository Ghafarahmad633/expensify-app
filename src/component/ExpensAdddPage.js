import React from "react";
import ExpenseForm from './ExpenseForm'
import {connect} from 'react-redux'
import {startAddExpense} from '../actions/expenses'
export class ExpensAdddPage extends React.Component{
    onSubmit=(expenses)=>{
        this.props.startAddExpense(expenses)
        this.props.history.push("/")
    }
    render(){
       return(
           <div>
               <div className='page-header'>
                   <div className='content-continer'>
                       <h1 className='page-header__title'>Add Expense</h1>
                   </div>
               </div>

               <div className='content-continer'>
                   <ExpenseForm onSubmit={this.onSubmit}/>
               </div>
           </div>
       )
    }
}
const mapDispatchToProps=(dispatch)=>{
    return{
        startAddExpense:(expenses)=>dispatch(startAddExpense(expenses))
    }
}
export default connect(undefined,mapDispatchToProps)(ExpensAdddPage)