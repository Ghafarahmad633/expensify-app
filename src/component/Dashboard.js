import React from "react";

import ExpenseList from "./ExpenseList";
import ExpenseListFilters from './ExpenseListFilters'
import TotalExpense from './TotalExpense'
class ExpensifyDashboardPage extends React.Component{

    render() {
        return (
            <div>
                <div>
                    <TotalExpense/>
                    <ExpenseListFilters/>
                    <ExpenseList />
                </div>
            </div>
        );
    }
}

export default ExpensifyDashboardPage