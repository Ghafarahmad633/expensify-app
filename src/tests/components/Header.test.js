import React from 'react'
import {shallow } from 'enzyme'
import {startLogOut,startLogin} from '../../actions/auth'

import {Header} from '../../component/Header'
test("Should Setup the Header Correctly",()=>{
    const wraper=shallow (<Header startLogOut={()=>{}}/>)
    expect(wraper).toMatchSnapshot()
    // expect(wraper.find('h1').text()).toBe("Expensifiy")    // const renderer=new RectShallowRenderer();
    // renderer.render(<Header/>)
    // expect(renderer.getRenderOutput()).toMatchSnapshot()
})


test("Should start logout n button Click",()=>{
    const startLogOut=jest.fn()
    const wraper=shallow(<Header startLogOut={startLogOut}/>)
    wraper.find('button').simulate('click')
    expect(startLogOut).toHaveBeenCalled()
})