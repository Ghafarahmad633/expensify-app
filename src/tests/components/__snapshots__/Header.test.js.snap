// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Should Setup the Header Correctly 1`] = `
<div
  className="header"
>
  <div
    className="content-continer"
  >
    <div
      className="header__content"
    >
      <Link
        className="header__title"
        replace={false}
        to="/dashboard"
      >
        <h1>
          Expensifiy
        </h1>
      </Link>
      <button
        className="button button__link"
        onClick={[Function]}
      >
        Logout
      </button>
    </div>
  </div>
</div>
`;
